refactor: remove resolvers, remove promise in withBeginTransaction, simplify code calling withBeginTransaction

danieljbruce opened the pull request:
**Changes:**

Modifies withBeginTransaction so that it doesn't await a promise anymore. Instead, it just calls the function that runs the code we want to run after ensuring the transaction has been started. Since the promise resolved to a UserCallbackData type, this type no longer needs to be used in this function because the promise isn't there anymore. Instead of relying on a value returned by withBeginTransaction and parsing that into callback arguments, withBeginTransaction just calls the callback directly which eliminates another layer in the call stack.

Code calling withBeginTransaction also doesn't use resolvers anymore and logic parsing the return value of withBeginTransaction isn't needed anymore either. This dramatically simplifies code necessary for all functions calling withBeginTransaction.

danieljbruce committed:
Change resolvers

Change the resolvers so that they all follow the same pattern.

danieljbruce committed:
Add a wrapWithBeginTransaction function for get

Generalize the calls to withBeginTransaction. Don’t allow the error type to be undefined in UserCallbackData since the error never actually will be undefined. This is necessary to solve various compile time errors.

danieljbruce committed:
Use wrapWithBeginTransaction in more places

commit and runQuery can now use #wrapWithBeginTransaction and this will reduce the amount of code necessary in each function respectively.

danieljbruce committed:
Add callbackWithError function

This function replaces a common pattern that occurs in transactions where we provide a callback that accepts an error as the first argument and some generic type that extends an array of any as the rest of the arguments.

danieljbruce committed:
Add return type to callbackWithError

Adding a return type just makes type checking stronger in the client library. We should introduce it here to make it more likely a compiler error will be thrown if the function is misused.

danieljbruce committed:
Delete code that is not used

When playing around with code to return a standard callback, code was written that is not required anymore. Also, rename StandardCallbackArgumentsAny.

danieljbruce committed:
There is no need to define a separate type

The UserCallbackArguments do not need a separate type if they are only used once. Define the type inline where it is needed.

danieljbruce committed:
TODO no longer applies

danieljbruce committed:
Remove TODO

danieljbruce committed:
Inline the error type

Error or null is only used in two places so inline it here.

danieljbruce committed:
Add a comment for callbackWithError

callbackWithError is an abstract function that needs to be explained. This comment helps with that.

danieljbruce committed:
Added comments to document sendUserCallbackData

sendUserCallbackData needs to be described so that its purpose is clear.

danieljbruce committed:
Move function

Move function to the right order alphabetically

danieljbruce committed:
Do not define a commit type only used once

Do not define a separate variable for only one use.

danieljbruce committed:
Replace the type in the comment with Function

Replace with Function to eliminate references to the generic parameter which make CI tests fail.

danieljbruce committed:
Just use function type for the callback

No need to mention the argument types and return types.

danieljbruce committed:
Remove withBeginTransaction

withBeginTransaction is only called by one function so remove it to condense call stack.

danieljbruce committed:
Simplify code usage of withBeginTransaction

Eliminate the need to build a resolver for all code usages of withBeginTransaction. Eliminate the need for a promise in withBeginTransaction. Eliminate the need for a callbackWithError function.

danieljbruce committed:
function should not have any arguments

danieljbruce committed:
Remove generic Args parameter

It is not used.

danieljbruce committed:
Remove TODOs that are not relevant anymore.

danieljbruce committed:
Rename to withBeginTransaction

Rename the function and eliminate redundant code.

danieljbruce committed:
Update comments for withBeginTransaction

Update the comments to reflect new parameters.

danieljbruce committed:
Simplify diff

danieljbruce committed:
Remove error as null

danieljbruce committed:
withBeginTransaction

Move withBeginTransaction to last alphabetically

daniel-sanche commented on src/transaction.ts:
```diff
@@ -1084,6 +1020,41 @@ class Transaction extends DatastoreRequest {
+            }
+          });
+        } catch (err: any) {
+          return callback(err);
```
What kind of errors might you encounter here? A comment would be useful

danieljbruce replied:
Added.

daniel-sanche commented on src/transaction.ts:
```diff
@@ -1084,6 +1020,41 @@ class Transaction extends DatastoreRequest {
+        try {
+          await this.#mutex.runExclusive(async () => {
+            if (this.#state === TransactionState.NOT_STARTED) {
+              const runResults = await this.#runAsync({gaxOptions});
```
nit: it looks like runAsync calls `beginTransaction`. Maybe name it something like `beginTransactionAsync`?

danieljbruce replied:
Sure, changed to beginTransactionAsync. Since this is in the transaction class run typically means run the transaction, but maybe beginTransactionAsync is more descriptive. Either way, this method is private so the name can be modified later if need be.

daniel-sanche reviewed (approved):
This is a great change! Just a couple comments

danieljbruce committed:
Add a comment to indicate error

In the code block it is not clear what kind of error will be produced and caught in catch so a comment is added to clarify the type of error that gets produced.

danieljbruce committed:
Rename method to beginTxAsync

Calling the method `run` is too ambiguous. Use #beginTransactionAsync instead.
